clients/client-go/tcobject/upload_download_test.go

```diff
@@ -0,0 +1,152 @@
+		MaxElapsedTime:      100 * time.Millisecond,
+		Clock:               backoff.SystemClock,
+	}
+	settings.Reset()
+	object.HTTPBackoffClient = &httpbackoff.Client{
+		BackOffSettings: settings,
+	}
+	return srv, r, object
+}
+
+// TestSimpleDownloadFails400 tests that when a simple download's GET fails
+// with a 400 HTTP status code, that a httpbackoff.BadHttpResponseCode error is
+// returned and the download isn't (non-intermittent status code).
+func TestSimpleDownloadFails400(t *testing.T) {
+	srv, r, object := mockObjectServer(t)
+	defer srv.Close()
+
+	r.HandleFunc("/simple",
+		http.HandlerFunc(
+			func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(400)
+				_, err := w.Write([]byte("uhoh"))
+				if err != nil {
+					t.Fatal(err)
+				}
+			},
+		),
+	).Methods("GET")
+
+	_, err := object.CreateUpload("some/object", &tcobject.CreateUploadRequest{})
+	assert.NoError(t, err)
+	err = object.FinishUpload("some/object", &tcobject.FinishUploadRequest{})
+	assert.NoError(t, err)
+	_, _, _, err = object.DownloadToBuf("some/object")
+
+	// make sure we got 400 status code for download URL
+	assert.IsType(t, tcobject.HTTPRetryError{}, err)
+	assert.IsType(t, httpbackoff.BadHttpResponseCode{}, err.(tcobject.HTTPRetryError).Err)
+	assert.Equal(t, 400, err.(tcobject.HTTPRetryError).Err.(httpbackoff.BadHttpResponseCode).HttpResponseCode)
+	assert.Equal(t, 1, err.(tcobject.HTTPRetryError).Attempts)
+}
+
+// TestSimpleDownloadFailsRetried tests that when a simple download's GET fails
+// with a 500 HTTP status code, that a httpbackoff.BadHttpResponseCode error is
+// returned and the download is retried twice (intermittent status code).
+func TestSimpleDownloadFailsRetried(t *testing.T) {
+	srv, r, object := mockObjectServer(t)
+	defer srv.Close()
+
+	r.HandleFunc("/simple",
+		http.HandlerFunc(
+			func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(500)
+				_, err := w.Write([]byte("uhoh"))
+				if err != nil {
+					t.Fatal(err)
+				}
+			},
+		),
+	).Methods("GET")
+
+	_, err := object.CreateUpload("some/object", &tcobject.CreateUploadRequest{})
+	assert.NoError(t, err)
+	err = object.FinishUpload("some/object", &tcobject.FinishUploadRequest{})
+	assert.NoError(t, err)
+	_, _, _, err = object.DownloadToBuf("some/object")
+
+	// make sure we get an exception
+	assert.IsType(t, tcobject.HTTPRetryError{}, err)
+	assert.IsType(t, httpbackoff.BadHttpResponseCode{}, err.(tcobject.HTTPRetryError).Err)
+	assert.Equal(t, 500, err.(tcobject.HTTPRetryError).Err.(httpbackoff.BadHttpResponseCode).HttpResponseCode)
+	assert.Equal(t, 3, err.(tcobject.HTTPRetryError).Attempts)
+}
+
+type testHandler struct {
+	mu       sync.Mutex
+	attempts uint8
+	t        *testing.T
+}
+
+func (th *testHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	th.mu.Lock()
+	defer th.mu.Unlock()
+	switch th.attempts {
+	case 0:
+		w.WriteHeader(500)
+		_, err := w.Write([]byte("uhoh"))
+		if err != nil {
+			th.t.Fatal(err)
```

Comment:
This can't be called from goroutines other than the main test goroutine.  Maybe use a channel?

https://golang.org/pkg/testing/#T
> A test ends when its Test function returns or calls any of the methods FailNow, Fatal, Fatalf, SkipNow, Skip, or Skipf. Those methods, as well as the Parallel method, must be called only from the goroutine running the Test function.

Reply:
I decided to remove the response body entirely, as it isn't needed for the test, and then it saves the more complex error handling.